adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/ErrorSubscription.java

```diff
@@ -0,0 +1,263 @@
+/**
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.adapter.mqtt;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.hono.auth.Device;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.EventConstants;
+import org.eclipse.hono.util.ResourceIdentifier;
+import org.eclipse.hono.util.Strings;
+import org.eclipse.hono.util.TelemetryConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.handler.codec.mqtt.MqttQoS;
+import io.vertx.mqtt.MqttTopicSubscription;
+
+/**
+ * A device's MQTT subscription for error messages.
+ * <p>
+ * Supported topic filters: {@code error|e/[TENANT]/[DEVICE_ID]/#}
+ * <br>(empty tenant/device segments can also be omitted)
+ * <p>
+ * Examples:
+ * <ol>
+ * <li>{@code error/DEFAULT_TENANT/4711/#} unauthenticated device</li>
+ * <li>{@code error/#} - authenticated device</li>
```

Comment:
based on the pattern above, wouldn't this rather be `error///#`? This would also better fit the pattern we use for subscribing to commands, wouldn't it?

Reply:
The minimal topic filter for command subscriptions is `command///req/#`, so there's still a non-empty segment after the empty tenant and device id parts.
I just think `error/#` is more intuitive and feels more natural than `error///#`. For an authenticated device it's just as short as using e.g. "event" for the other direction.
Of course `error/#` must not be interpreted as "get all error messages for *all* tenants/devices". But I don't think that this would be the intuitive expectation anyway since for authenticated (non-gateway) devices everything is done in the context of the device derived from the authentication data.
And users still can use `error///#` if they prefer.

Reply:
Is it a supported use case that a gateway device subscribes to errors for all devices that it may act on behalf of? If so, what is the topic that it would need to subscribe to? `error//+/#`?

Reply:
Yes. The possible topic filters for such a case are `error/#`, `error//+/#` and `error/${tenant-id}/+/#` (see also the "Receiving Error Messages (authenticated Gateway)" chapter in `mqtt-adapter.md`).
The gateway would otherwise use `error//${device-id}/#` or `error/${tenant-id}/${device-id}#` to get error messages for a specific device it acts on behalf of.

Reply:
So in these cases we follow MQTT wildcard semantics. However, `error/#` does not, right? FMPOV that is not very intuitive but confusing ...

Reply:
I don't see what you mean. `error/#` with the multi-level # wildcard would match all error topic strings that the device may receive messages for, meaning the # matches the tenant, device and following topic levels.

Reply:
Well, with the command subscription topics we deliberately made the change from `command/+/+/req/#` to `command///req/#` because the `+` wildcard character suggests that *any* tenant and device ID would match. FMPOV the same argument applies here as well, doesn't it?

Reply:
Well, the topic with empty tenant/device id segments are special anyway. in the sense that they make it clear that there is no single big, tenant/device-independent topic tree to tap into with the subscription topic filter. The `command///req/[*]/[*]` publish topic is used *across* devices.

For example, if a person manages 2 different (non-gateway) devices "deviceA" and "deviceB" and uses `command///req/#` for both their subscriptions, the observation to make is that both devices get their commands with topics like `command///req/111/setBrightness`. 
But, that person will notice that "deviceA" will only get the  `command///req/[*]/[*]` messages for "deviceA", although command messages for "deviceB" also fit the `command///req/#` pattern.

So a user has to understand that the messages received via a subscription are always just the ones for the authenticated device (or the gateway-managed devices).

In that sense, the understanding that `error/#` will also just match the authenticated device or gateway-managed devices isn't that unintuitive.

I see that mandating `error///#` will make it just a bit more obvious, what kind of topics a subscription of authenticated "deviceA" of "tenantX" will get (in the sense that it's necessarily `error///something` instead of potentially `error/tenantX/deviceA/something`). But I still think it still looks a bit odd to use `error///#` instead of `error/#` (with the realization in mind that the received messages are always device-specific anyway).

Reply:
Based on your arguments we could also simply let devices subscribe to `command/#` in order to receive commands, couldn't we? The problem with this approach is FMPOV that we start introducing *special cases* that we need to handle differently than usual. We deliberately chose to use the common addressing scheme *endpoint/tenant/deviceId/...* throughout our APIs. I do not see any convincing reason to divert from it now, in particular since nobody seems to have a problem with subscribing to `command///req/#` ...

Reply:
Ok, to keep it in accordance to the command pattern, I've made the corresponding changes so that `error///#` needs to be used now and `error/#` isn't supported.